From 20b8089401a9845847d21534fd005dc4d094cd71 Mon Sep 17 00:00:00 2001
From: Kim Eik <kim@heldig.org>
Date: Mon, 14 May 2012 14:11:13 +0200
Subject: [PATCH 31/32] added a optional maxzoom and minzoom parameter

this parameter will limit how far in and out of a map
a user can zoom.

as a side note, openlayers api for limiting zoom is horrible....

* Replaced default '' with false.
* Improved js for openlayers in regards of type checking.

Change-Id: I8ec1c2c56a55df5694d4e49f503947e06a1dec36
---
 includes/parserHooks/Maps_DisplayLine.php         |   15 +++++++++++
 includes/services/GoogleMaps3/jquery.googlemap.js |    5 ++++
 includes/services/OpenLayers/jquery.openlayers.js |   28 +++++++++++++++++++++
 3 files changed, 48 insertions(+)

diff --git a/includes/parserHooks/Maps_DisplayLine.php b/includes/parserHooks/Maps_DisplayLine.php
index 881411b..168232f 100644
--- a/includes/parserHooks/Maps_DisplayLine.php
+++ b/includes/parserHooks/Maps_DisplayLine.php
@@ -111,6 +111,21 @@ class MapsDisplayLine extends MapsDisplayPoint {
 		$params['static']->setDefault( false );
 		$params['static']->setDoManipulationOfDefault( false );
 
+		$params['maxzoom'] = new Parameter(
+			'maxzoom',
+			Parameter::TYPE_INTEGER
+		);
+		$params['maxzoom']->setDefault( false );
+		$params['maxzoom']->setDoManipulationOfDefault( false );
+		$params['maxzoom']->addDependencies( 'minzoom' );
+
+		$params['minzoom'] = new Parameter(
+			'minzoom',
+			Parameter::TYPE_INTEGER
+		);
+		$params['minzoom']->setDefault( false );
+		$params['minzoom']->setDoManipulationOfDefault( false );
+
 		return $params;
 	}
 
diff --git a/includes/services/GoogleMaps3/jquery.googlemap.js b/includes/services/GoogleMaps3/jquery.googlemap.js
index 0ac5c7e..d6bf979 100644
--- a/includes/services/GoogleMaps3/jquery.googlemap.js
+++ b/includes/services/GoogleMaps3/jquery.googlemap.js
@@ -342,6 +342,11 @@
 				mapTypeIds:options.types
 			};
 
+
+			//max/min -zoom
+			mapOptions.maxZoom = options.maxzoom === false ? null : options.maxzoom;
+			mapOptions.minZoom = options.minzoom === false ? null : options.minzoom;
+
 			//static mode
 			if (options['static']) {
 				mapOptions.draggable = false;
diff --git a/includes/services/OpenLayers/jquery.openlayers.js b/includes/services/OpenLayers/jquery.openlayers.js
index 46e9e8e..0531ff1 100644
--- a/includes/services/OpenLayers/jquery.openlayers.js
+++ b/includes/services/OpenLayers/jquery.openlayers.js
@@ -272,6 +272,34 @@
 			this.addControls(map, options.controls, this.get(0));
 		}
 
+        //ugly hack to allow for min/max zoom
+        if (options.maxzoom !== false || options.minzoom !== false) {
+
+            if(options.maxzoom === false){
+                options.maxzoom = mapOptions.numZoomLevels;
+            }
+
+            map.getNumZoomLevels = function () {
+	            var zoomLevels = 1;
+	            zoomLevels += options.maxzoom !== false ? options.maxzoom : 0;
+	            zoomLevels -= options.minzoom !== false ? options.minzoom : 0;
+                return zoomLevels;
+            };
+
+            map.isValidZoomLevel = function (zoomLevel) {
+                var valid = ( (zoomLevel != null) &&
+                    (zoomLevel >= options.minzoom) &&
+                    (zoomLevel <= options.maxzoom) );
+                if (!valid && map.getZoom() == 0) {
+	                var maxzoom = options.maxzoom !== false ? options.maxzoom : 0;
+	                var minzoom = options.minzoom !== false ? options.minzoom : 0;
+	                var zoom =  maxzoom - (maxzoom - minzoom) / 2
+                    map.zoomTo(zoom);
+                }
+                return valid;
+            }
+        }
+
 		// Add the base layers.
 		for (i = 0, n = options.layers.length; i < n; i++) {
 			map.addLayer(options.layers[i]);
-- 
1.7.9.5

